### DISCUSSION TITLE
Notarisation in macOS Sonoma

### DISCUSSION THREAD

**Discussion by johnalanwoods** [Homebrew/discussions#4582] [Getting started] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.21
ORIGIN: https://github.com/Homebrew/brew
HEAD: 9d0b7361d1a197949dbb47b761405075f83983b7
Last commit: 2 weeks ago
Core tap JSON: 08 Jun 16:43 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.1
Rosetta 2: false
```
```


### Output of `brew doctor`

```shell
Ready to brew
```

### Description of issue

Will the requirements for Notarisation in macOS Sonoma and beyond affect Homebrew?

Codesigning can be done locally, but Notarisation is a separate thing.

It seems Notarisation is required going forward: https://developer.apple.com/documentation/security/notarizing_macos_software_before_distribution

How will thing affect Homebrew, will local tools be possible to run without Notarisation/checking in with Apple's servers on each run?

**Comment by p-linnane** [comment]
We are already requiring new casks to be notarized. At some point I'm sure we will require it for all casks, but that's not something we have discussed internally yet.

You may be interested in this great post from @hoakleyelc on his findings for changes to notarization in Sonoma: https://eclecticlight.co/2023/06/18/last-week-on-my-mac-lies-notarization-and-privacy/

**Reply by johnalanwoods** [reply]
Hey thanks for the reply @p-linnane - great article as always by Howard.

Something that confuses me though, the difference between:

```macOS requires that executables are signed (at least on Apple silicon Macs), but that can be ad hoc, and doesn’t require an Apple signing certificate.```

and 

```Apps and other executable code distributed by developers must be fully signed and notarized, but normally there’s no requirement for sandboxing.```

I don't understand the difference, can't the former be "distributed"?

Also, what is done for bottles atm (I mean the CLI apps that are precompiled), do you notarise that, or maybe it doesn't matter for command line?

**Reply by SMillerDev** [reply, marked as answer]
> I don't understand the difference, can't the former be "distributed"?

They can not, ad hoc only works on that one device.

> Also, what is done for bottles atm (I mean the CLI apps that are precompiled), do you notarise that, or maybe it doesn't matter for command line?

Bottled apps are ad hoc signed on your device upon install

**Reply by Bo98** [reply]
> You can notarize several different types of software deliverables, including:
> * macOS apps
> * Non-app bundles, such as kernel extensions
> * Disk images (UDIF format)
> * Flat installer packages

For formulae (which are usually CLI executables), notarisation does not apply. The definition of an "app" here is exclusively the `.app` bundles you would typically have in `/Applications`. I'm not aware of any changes to notarisation in Sonoma - it's been a requirement for newly built apps since 10.15.

The only requirement for regular executables is the code signing on Apple Silicon, and that can be done ad-hoc (which we do).

**Reply by johnalanwoods** [reply]
> For formulae (which are usually CLI executables), notarisation does not apply.

Why is this the case?

**Reply by gromgit** [reply]
Because Homebrew distributes binaries with relocatable paths, and the binary editing that is done during the actual install makes pre-signing binaries useless, much less notarization.

**Comment by johnalanwoods** [comment]
OK cool, so basically Bottles compiled by Homebrew or indeed src that compiles locally via Homebrew is not notarised and won't be for the foreseeable future, instead it's only adhoc signed... does this affect execution time checking in with Apple's servers when running an executable?

> They can not, ad hoc only works on that one device.
I wonder how this works, maybe includes a device UUID or equiv.

Thanks all for your help.
